examples/silentpayments.c

```diff
@@ -0,0 +1,417 @@
+         * to do an expensive pubkey generation. This is not needed for plain seckeys since there is no need
+         * for negation.
+         *
+         * The public key from each input keypair is saved in the `tx_inputs` array. This array will be used
+         * later in the example to represent the public keys the recipient will extracte from the
+         * transaction inputs.
+         */
+
+        for (i = 0; i < 2; i++) {
+            /* If the secret key is zero or out of range (bigger than secp256k1's
+             * order), we try to sample a new key. Note that the probability of this
+             * happening is negligible. */
+            while (1) {
+                unsigned char seckey[32];
+                if (!fill_random(seckey, sizeof(seckey))) {
+                    printf("Failed to generate randomness\n");
+                    return 1;
+                }
+                /* Try to create a keypair with a valid context, it should only fail if
+                 * the secret key is zero or out of range. */
+                if (secp256k1_keypair_create(ctx, &sender_seckeys[i], seckey)) {
+                    sender_seckey_ptrs[i] = &sender_seckeys[i];
+                    ret = secp256k1_keypair_xonly_pub(ctx, &tx_inputs[i], NULL, &sender_seckeys[i]);
+                    assert(ret);
+                    break;
+                } else {
+                    printf("Failed to create keypair\n");
+                    return 1;
+                }
+            }
+        }
+        /*** Create the recipient objects ***/
+
+        /* Alice is sending to Bob and Carol in this transaction:
+         *
+         *     1. One output to Bob's labelled address
+         *     2. Two outputs for Carol (1.0 and 3.0 bitcoin)
+         *
+         * Alice creates the recipient objects and adds the index of the original ordering (i.e. the ordering
+         * of the `sp_addresses` array) to each object. This index is used to return the generated outputs in
+         * the original ordering so that Alice can match up the generated outputs with the correct amounts.
+         *
+         * Note: to create multiple outputs for Carol, Alice simply passes her silent payment
+         * address mutltiple times.
+         */
+        sp_addresses[0] = &carol_address; /* : 1.0 BTC */
+        sp_addresses[1] = &bob_address;   /* : 2.0 BTC */
+        sp_addresses[2] = &carol_address; /* : 3.0 BTC */
+        for (i = 0; i < N_TX_OUTPUTS; i++) {
+            ret = secp256k1_ec_pubkey_parse(ctx, &recipients[i].scan_pubkey, (*(sp_addresses[i]))[0], 33);
+            assert(ret);
+            ret = secp256k1_ec_pubkey_parse(ctx, &recipients[i].spend_pubkey, (*(sp_addresses[i]))[1], 33);
+            assert(ret);
+            recipients[i].index = i;
+            recipient_ptrs[i] = &recipients[i];
+        }
+        for (i = 0; i < N_TX_OUTPUTS; i++) {
+            generated_output_ptrs[i] = &tx_outputs[i];
```

Comment:
Good catch, I think this would be less confusing if more consistent naming were used, e.g., `tx_outputs; tx_output_ptrs`. Regarding the API: arrays are passed as a pointer to an array of pointers to objects. This means we need to first create the array of objects (`tx_outputs`), and then create the array of pointers to those objects (`tx_output_ptrs`).

In the case of `tx_outputs`, its less obvious why the we want the pointer to array of pointers approach, but I think its better to keep the API consistent, i.e., if passing an array to the function, it must be a pointer to an array of pointers.

I think this might warrant a `doc/silentpayments.md` explaining the usage of the API, similar to the musig2 PR.

Reply:
Agree it's better to keep the API consistent.

There's no magical C way to create an array of pointers to objects without first going through an array of objects? In any case the naming change makes it clear that `tx_outputs` and `tx_output_ptrs` refer to the same "stuff".